crud/common/schema.lua

```diff
@@ -0,0 +1,107 @@
+local fiber = require('fiber')
+local msgpack = require('msgpack')
+local digest = require('digest')
+local vshard = require('vshard')
+
+local const = require('crud.common.const')
+
+local schema = {}
+
+local function call_reload_schema(replicasets)
+    for _, replicaset in pairs(replicasets) do
+        local master = replicaset.master
+        master.conn:reload_schema()
+    end
+end
+
+local reload_in_progress = false
+local reload_schema_cond = fiber.cond()
+
+local function reload_schema(replicasets)
+    if reload_in_progress then
+        reload_schema_cond:wait()
+    else
+        reload_in_progress = true
+
+        call_reload_schema(replicasets)
+
+        reload_schema_cond:broadcast()
+        reload_in_progress = false
+    end
+
+    return true
+end
+
+function schema.wrap_func_reload(func, ...)
+    local i = 0
+
+    local res, err, need_reload
+    while true do
+        res, err, need_reload = func(...)
+
+        if err == nil or not need_reload then
+            break
+        end
+
+        reload_schema(vshard.router.routeall())
+
+        i = i + 1
+        if i > const.RELOAD_RETRIES_NUM then
+            break
+        end
+    end
+
+    return res, err
+end
+
+local function get_space_schema_hash(space)
+    if space == nil then
+        return ''
+    end
```

Comment:
Are there some cases when `space == nil`? Looks like a mistake.

Reply:
Yes, there is not such cases, but this is just a function that computes space format hash and it knows nothing about this cases. Without this check we will got error on using `space.index` sometimes.

Reply:
> Yes, there are no such cases

You need a put assert/error here

Reply:
In fact, it's a valid case - when space doesn't exist. Why do we need error here?

Reply:
If it's ok, why do you return empty string but not e.g. nil?

Reply:
But what's the difference?)

Reply:
String even empty is a content that could be useful for some cases/reasons. `nil` is absence of anything. Just value that says you "I don't have nothing useful".

Reply:
See `schema.result_needs_reload` function. `nil` returned by storage means "I didn't returned any hash, you have nothing to compare with", but `""` means "My space schema is that: space doesn't exist".